File: fons/norma-rs/hal/arca.rs
//! arca.rs - Database Device Implementation
//!
//! Native Rust implementation of the HAL database interface.
//! Uses sqlx with multi-database support (SQLite, PostgreSQL, MySQL).
//!
//! Verb conjugation encodes async and cardinality:
//!   - Future singular (-et): async, returns one value
//!   - Future plural (-ent): async generator, yields multiple values

use async_stream::stream;
use futures::Stream;
use serde_json::Value;
use sqlx::any::{AnyPoolOptions, AnyRow};
use sqlx::{AnyPool, Column, Row, TypeInfo};
use std::pin::Pin;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

/// Database connection wrapper
pub struct Connexio {
    pool: AnyPool,
    open: Arc<AtomicBool>,
}

/// Database transaction wrapper
pub struct Transactio {
    tx: Option<sqlx::Transaction<'static, sqlx::Any>>,
}

// =============================================================================
// CONNECTION
// =============================================================================

/// Connect to database (driver inferred from URL scheme)
/// URLs: postgres://, mysql://, sqlite:///path, sqlite://:memory:
pub async fn connectet(url: &str) -> Connexio {
    // Install default drivers
    sqlx::any::install_default_drivers();

    let pool = AnyPoolOptions::new()
        .max_connections(5)
        .connect(url)
        .await
        .expect("failed to connect to database");

    Connexio {
        pool,
        open: Arc::new(AtomicBool::new(true)),
    }
}

impl Connexio {
    // =========================================================================
    // QUERIES
    // =========================================================================

    /// Stream rows as async generator
    pub fn quaerent(
        &self,
        sql: &str,
        params: &[Value],
    ) -> Pin<Box<dyn Stream<Item = Value> + Send + '_>> {
        let sql = sql.to_string();
        let params = params.to_vec();
        let pool = self.pool.clone();

        Box::pin(stream! {
            let query = build_query(&sql, &params);
            let mut rows = query.fetch(&pool);

            use futures::StreamExt;
            while let Some(row) = rows.next().await {
                if let Ok(row) = row {
                    yield row_to_value(&row);
                }
            }
        })
    }

    /// Return all rows as list
    pub async fn quaeret(&self, sql: &str, params: &[Value]) -> Vec<Value> {
        let query = build_query(sql, params);
        let rows = query
            .fetch_all(&self.pool)
            .await
            .expect("failed to execute query");

        rows.iter().map(row_to_value).collect()
    }

    /// Return first row or None
    pub async fn capiet(&self, sql: &str, params: &[Value]) -> Option<Value> {
        let query = build_query(sql, params);
        let row = query.fetch_optional(&self.pool).await.expect("failed to execute query");

        row.as_ref().map(row_to_value)
    }

    // =========================================================================
    // MUTATIONS
    // =========================================================================

    /// Execute INSERT/UPDATE/DELETE, return affected row count
    pub async fn exsequetur(&self, sql: &str, params: &[Value]) -> u64 {
        let query = build_query(sql, params);
        let result = query
            .execute(&self.pool)
            .await
            .expect("failed to execute mutation");

        result.rows_affected()
    }

    /// Execute INSERT, return last inserted ID
    pub async fn inseret(&self, sql: &str, params: &[Value]) -> i64 {
        let query = build_query(sql, params);
        let result = query
            .execute(&self.pool)
            .await
            .expect("failed to execute insert");

        result.last_insert_id().unwrap_or(0)
    }

    // =========================================================================
    // TRANSACTIONS
    // =========================================================================

    /// Begin transaction
    pub async fn incipiet(&self) -> Transactio {
        let tx = self
            .pool
            .begin()
            .await
            .expect("failed to begin transaction");

        Transactio { tx: Some(tx) }
    }

    // =========================================================================
    // LIFECYCLE
    // =========================================================================

    /// Close connection
    pub fn claude(&self) {
        self.open.store(false, Ordering::SeqCst);
        // Pool will be dropped when Connexio is dropped
    }

    /// Check if connection is open
    pub fn aperta(&self) -> bool {
        self.open.load(Ordering::SeqCst)
    }
}

impl Transactio {
    /// Stream rows as async generator within transaction
    pub fn quaerent(
        &mut self,
        sql: &str,
        params: &[Value],
    ) -> Pin<Box<dyn Stream<Item = Value> + Send + '_>> {
        let sql = sql.to_string();
        let params = params.to_vec();

        // Get mutable reference to transaction
        let tx = self.tx.as_mut().expect("transaction already consumed");

        Box::pin(stream! {
            let query = build_query(&sql, &params);
            let mut rows = query.fetch(&mut **tx);

            use futures::StreamExt;
            while let Some(row) = rows.next().await {
                if let Ok(row) = row {
                    yield row_to_value(&row);
                }
            }
        })
    }

    /// Return all rows as list within transaction
    pub async fn quaeret(&mut self, sql: &str, params: &[Value]) -> Vec<Value> {
        let tx = self.tx.as_mut().expect("transaction already consumed");
        let query = build_query(sql, params);
        let rows = query
            .fetch_all(&mut **tx)
            .await
            .expect("failed to execute query");

        rows.iter().map(row_to_value).collect()
    }

    /// Execute mutation within transaction
    pub async fn exsequetur(&mut self, sql: &str, params: &[Value]) -> u64 {
        let tx = self.tx.as_mut().expect("transaction already consumed");
        let query = build_query(sql, params);
        let result = query
            .execute(&mut **tx)
            .await
            .expect("failed to execute mutation");

        result.rows_affected()
    }

    /// Commit transaction
    pub async fn committet(mut self) {
        let tx = self.tx.take().expect("transaction already consumed");
        tx.commit().await.expect("failed to commit transaction");
    }

    /// Rollback transaction
    pub async fn revertet(mut self) {
        let tx = self.tx.take().expect("transaction already consumed");
        tx.rollback().await.expect("failed to rollback transaction");
    }
}

// =============================================================================
// HELPERS
// =============================================================================

/// Build a query with bound parameters
fn build_query<'a>(
    sql: &'a str,
    params: &'a [Value],
) -> sqlx::query::Query<'a, sqlx::Any, sqlx::any::AnyArguments<'a>> {
    let mut query = sqlx::query(sql);

    for param in params {
        query = match param {
            Value::Null => query.bind(None::<String>),
            Value::Bool(b) => query.bind(*b),
            Value::Number(n) => {
                if let Some(i) = n.as_i64() {
                    query.bind(i)
                } else if let Some(f) = n.as_f64() {
                    query.bind(f)
                } else {
                    query.bind(None::<String>)
                }
            }
            Value::String(s) => query.bind(s.as_str()),
            Value::Array(_) | Value::Object(_) => {
                // Serialize complex types as JSON strings
                query.bind(serde_json::to_string(param).unwrap_or_default())
            }
        };
    }

    query
}

/// Convert a database row to a JSON value (object with column names as keys)
fn row_to_value(row: &AnyRow) -> Value {
    let mut obj = serde_json::Map::new();

    for col in row.columns() {
        let name = col.name().to_string();
        let type_name = col.type_info().name();

        let value: Value = match type_name {
            "INTEGER" | "INT" | "INT4" | "INT8" | "BIGINT" | "SMALLINT" => {
                row.try_get::<i64, _>(col.ordinal())
                    .map(Value::from)
                    .unwrap_or(Value::Null)
            }
            "REAL" | "FLOAT" | "FLOAT4" | "FLOAT8" | "DOUBLE" | "NUMERIC" | "DECIMAL" => {
                row.try_get::<f64, _>(col.ordinal())
                    .map(Value::from)
                    .unwrap_or(Value::Null)
            }
            "BOOLEAN" | "BOOL" => row
                .try_get::<bool, _>(col.ordinal())
                .map(Value::from)
                .unwrap_or(Value::Null),
            "TEXT" | "VARCHAR" | "CHAR" | "STRING" => row
                .try_get::<String, _>(col.ordinal())
                .map(Value::from)
                .unwrap_or(Value::Null),
            "BLOB" | "BYTEA" => row
                .try_get::<Vec<u8>, _>(col.ordinal())
                .map(|bytes| Value::from(base64_encode(&bytes)))
                .unwrap_or(Value::Null),
            _ => {
                // Default: try as string
                row.try_get::<String, _>(col.ordinal())
                    .map(Value::from)
                    .unwrap_or(Value::Null)
            }
        };

        obj.insert(name, value);
    }

    Value::Object(obj)
}

/// Simple base64 encoding for BLOB fields
fn base64_encode(data: &[u8]) -> String {
    use std::io::Write;
    let mut buf = Vec::new();
    let mut encoder = Base64Encoder::new(&mut buf);
    encoder.write_all(data).unwrap();
    drop(encoder);
    String::from_utf8(buf).unwrap_or_default()
}

/// Minimal base64 encoder (avoiding extra dependencies)
struct Base64Encoder<W: std::io::Write> {
    writer: W,
}

impl<W: std::io::Write> Base64Encoder<W> {
    fn new(writer: W) -> Self {
        Self { writer }
    }
}

impl<W: std::io::Write> std::io::Write for Base64Encoder<W> {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        const ALPHABET: &[u8; 64] =
            b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

        for chunk in buf.chunks(3) {
            let mut n = (chunk[0] as u32) << 16;
            if chunk.len() > 1 {
                n |= (chunk[1] as u32) << 8;
            }
            if chunk.len() > 2 {
                n |= chunk[2] as u32;
            }

            self.writer.write_all(&[ALPHABET[(n >> 18 & 63) as usize]])?;
            self.writer.write_all(&[ALPHABET[(n >> 12 & 63) as usize]])?;

            if chunk.len() > 1 {
                self.writer.write_all(&[ALPHABET[(n >> 6 & 63) as usize]])?;
            } else {
                self.writer.write_all(b"=")?;
            }

            if chunk.len() > 2 {
                self.writer.write_all(&[ALPHABET[(n & 63) as usize]])?;
            } else {
                self.writer.write_all(b"=")?;
            }
        }
        Ok(buf.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        self.writer.flush()
    }
}

File: runtimes/norma-py/hal/arca.py
"""arca.py - Database Device Implementation

Native Python implementation of the HAL database interface.
Uses sqlite3 for SQLite, and optional asyncpg/aiomysql for PostgreSQL/MySQL.

Verb conjugation encodes async and cardinality:
  - Future singular (-et): async, returns one value
  - Future plural (-ent): async generator, yields multiple values
"""

import sqlite3
from dataclasses import dataclass, field
from typing import Any, Optional, AsyncIterator
from urllib.parse import urlparse

try:
    import asyncpg
    _HAS_ASYNCPG = True
except ImportError:
    _HAS_ASYNCPG = False

try:
    import aiomysql
    _HAS_AIOMYSQL = True
except ImportError:
    _HAS_AIOMYSQL = False


@dataclass
class Connexio:
    """Database connection wrapper."""
    _conn: Any
    _driver: str
    _open: bool = field(default=True)

    # =========================================================================
    # QUERIES
    # =========================================================================

    async def quaerent(self, sql: str, params: list[Any]) -> AsyncIterator[dict[str, Any]]:
        """Stream rows as async generator."""
        if self._driver == "sqlite":
            cursor = self._conn.execute(sql, params)
            columns = [desc[0] for desc in cursor.description]
            for row in cursor:
                yield dict(zip(columns, row))
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            async with self._conn.transaction():
                async for record in self._conn.cursor(sql, *params):
                    yield dict(record)
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            async with self._conn.cursor(aiomysql.DictCursor) as cur:
                await cur.execute(sql, params)
                async for row in cur:
                    yield row

    async def quaeret(self, sql: str, params: list[Any]) -> list[dict[str, Any]]:
        """Return all rows as list."""
        if self._driver == "sqlite":
            cursor = self._conn.execute(sql, params)
            columns = [desc[0] for desc in cursor.description]
            return [dict(zip(columns, row)) for row in cursor.fetchall()]
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            rows = await self._conn.fetch(sql, *params)
            return [dict(row) for row in rows]
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            async with self._conn.cursor(aiomysql.DictCursor) as cur:
                await cur.execute(sql, params)
                return await cur.fetchall()
        return []

    async def capiet(self, sql: str, params: list[Any]) -> Optional[dict[str, Any]]:
        """Return first row or None."""
        if self._driver == "sqlite":
            cursor = self._conn.execute(sql, params)
            columns = [desc[0] for desc in cursor.description]
            row = cursor.fetchone()
            return dict(zip(columns, row)) if row else None
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            row = await self._conn.fetchrow(sql, *params)
            return dict(row) if row else None
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            async with self._conn.cursor(aiomysql.DictCursor) as cur:
                await cur.execute(sql, params)
                return await cur.fetchone()
        return None

    # =========================================================================
    # MUTATIONS
    # =========================================================================

    async def exsequetur(self, sql: str, params: list[Any]) -> int:
        """Execute INSERT/UPDATE/DELETE, return affected row count."""
        if self._driver == "sqlite":
            cursor = self._conn.execute(sql, params)
            self._conn.commit()
            return cursor.rowcount
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            result = await self._conn.execute(sql, *params)
            return int(result.split()[-1]) if result else 0
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            async with self._conn.cursor() as cur:
                await cur.execute(sql, params)
                await self._conn.commit()
                return cur.rowcount
        return 0

    async def inseret(self, sql: str, params: list[Any]) -> int:
        """Execute INSERT, return last inserted ID."""
        if self._driver == "sqlite":
            cursor = self._conn.execute(sql, params)
            self._conn.commit()
            return cursor.lastrowid or 0
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            row = await self._conn.fetchrow(sql + " RETURNING id", *params)
            return row["id"] if row else 0
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            async with self._conn.cursor() as cur:
                await cur.execute(sql, params)
                await self._conn.commit()
                return cur.lastrowid or 0
        return 0

    # =========================================================================
    # TRANSACTIONS
    # =========================================================================

    async def incipiet(self) -> "Transactio":
        """Begin transaction."""
        if self._driver == "sqlite":
            return Transactio(_conn=self._conn, _driver=self._driver)
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            tx = self._conn.transaction()
            await tx.start()
            return Transactio(_conn=self._conn, _driver=self._driver, _tx=tx)
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            await self._conn.begin()
            return Transactio(_conn=self._conn, _driver=self._driver)
        return Transactio(_conn=self._conn, _driver=self._driver)

    # =========================================================================
    # LIFECYCLE
    # =========================================================================

    def claude(self) -> None:
        """Close connection."""
        self._open = False
        if self._driver == "sqlite":
            self._conn.close()

    def aperta(self) -> bool:
        """Check if connection is open."""
        return self._open


@dataclass
class Transactio:
    """Database transaction wrapper."""
    _conn: Any
    _driver: str
    _tx: Any = None

    async def quaerent(self, sql: str, params: list[Any]) -> AsyncIterator[dict[str, Any]]:
        """Stream rows as async generator within transaction."""
        if self._driver == "sqlite":
            cursor = self._conn.execute(sql, params)
            columns = [desc[0] for desc in cursor.description]
            for row in cursor:
                yield dict(zip(columns, row))
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            async for record in self._conn.cursor(sql, *params):
                yield dict(record)
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            async with self._conn.cursor(aiomysql.DictCursor) as cur:
                await cur.execute(sql, params)
                async for row in cur:
                    yield row

    async def quaeret(self, sql: str, params: list[Any]) -> list[dict[str, Any]]:
        """Return all rows as list within transaction."""
        if self._driver == "sqlite":
            cursor = self._conn.execute(sql, params)
            columns = [desc[0] for desc in cursor.description]
            return [dict(zip(columns, row)) for row in cursor.fetchall()]
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            rows = await self._conn.fetch(sql, *params)
            return [dict(row) for row in rows]
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            async with self._conn.cursor(aiomysql.DictCursor) as cur:
                await cur.execute(sql, params)
                return await cur.fetchall()
        return []

    async def exsequetur(self, sql: str, params: list[Any]) -> int:
        """Execute mutation within transaction."""
        if self._driver == "sqlite":
            cursor = self._conn.execute(sql, params)
            return cursor.rowcount
        elif self._driver == "postgres" and _HAS_ASYNCPG:
            result = await self._conn.execute(sql, *params)
            return int(result.split()[-1]) if result else 0
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            async with self._conn.cursor() as cur:
                await cur.execute(sql, params)
                return cur.rowcount
        return 0

    async def committet(self) -> None:
        """Commit transaction."""
        if self._driver == "sqlite":
            self._conn.commit()
        elif self._driver == "postgres" and _HAS_ASYNCPG and self._tx:
            await self._tx.commit()
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            await self._conn.commit()

    async def revertet(self) -> None:
        """Rollback transaction."""
        if self._driver == "sqlite":
            self._conn.rollback()
        elif self._driver == "postgres" and _HAS_ASYNCPG and self._tx:
            await self._tx.rollback()
        elif self._driver == "mysql" and _HAS_AIOMYSQL:
            await self._conn.rollback()


# =============================================================================
# CONNECTION
# =============================================================================

async def connectet(url: str) -> Connexio:
    """Connect to database (driver inferred from URL scheme).

    URLs: postgres://, mysql://, sqlite:///path, sqlite://:memory:
    """
    parsed = urlparse(url)
    scheme = parsed.scheme

    if scheme == "sqlite":
        path = parsed.path
        if path.startswith("//"):
            path = path[2:]
        conn = sqlite3.connect(path)
        return Connexio(_conn=conn, _driver="sqlite")

    elif scheme in ("postgres", "postgresql"):
        if not _HAS_ASYNCPG:
            raise RuntimeError("asyncpg package required for PostgreSQL")
        conn = await asyncpg.connect(url)
        return Connexio(_conn=conn, _driver="postgres")

    elif scheme == "mysql":
        if not _HAS_AIOMYSQL:
            raise RuntimeError("aiomysql package required for MySQL")
        conn = await aiomysql.connect(
            host=parsed.hostname or "localhost",
            port=parsed.port or 3306,
            user=parsed.username or "",
            password=parsed.password or "",
            db=parsed.path.lstrip("/"),
        )
        return Connexio(_conn=conn, _driver="mysql")

    else:
        raise ValueError(f"Unsupported database scheme: {scheme}")
